### PR TITLE
set exchange's fromEndpoint property through the UnitOfWork advice

### PR DESCRIPTION
**Pull request by bvahdat**

### CODE REVIEW

File: core/camel-base-engine/src/main/java/org/apache/camel/impl/engine/CamelInternalProcessor.java [hunk 1]
```diff
@@ -731,6 +731,7 @@ public UnitOfWork before(Exchange exchange) throws Exception {
                     this.routeId = route.getRouteId();
                 }
                 ExtendedExchange ee = (ExtendedExchange) exchange;
+                ee.setFromEndpoint(route.getEndpoint());
```
Comment (davsclaus) [inline core/camel-base-engine/src/main/java/org/apache/camel/impl/engine/CamelInternalProcessor.java:734]: Whats the use case for this? 

The exchange could already have been assigned with an endpoint, so you would potentially override this.

**Comment by bvahdat** [comment]
The idea is to avoid custom logic such as the following when a `org.apache.camel.support.AbstractExchange` object is created using a parent one:

```
String fromEndpointUri = null;
if (message.getExchange().getFromEndpoint() != null) {
    fromEndpointUri = message.getExchange().getFromEndpoint().getEndpointUri();
} else {
    Route fromRoute = message.getExchange().getContext().getRoute(message.getExchange().getFromRouteId());
    if (fromRoute != null) {
        fromEndpointUri = fromRoute.getEndpoint().getEndpointUri();
    }
}
```

File: core/camel-base-engine/src/main/java/org/apache/camel/impl/engine/CamelInternalProcessor.java (hunk 1 above)
Comment (bvahdat) [reply core/camel-base-engine/src/main/java/org/apache/camel/impl/engine/CamelInternalProcessor.java:734]: Thanks for your feedback. I was just writing the comment below at the same time as you commented on the change here.

**Comment by davsclaus** [comment]
Okay I did a bit of test, and we have tests from camel-core that would create an exchange from another endpoint (such as file endpoint) and then send to direct endpoint via a producer template to be routed.

In that situation then your patch will override the existing from endpoint (the file endpoint) with the route endpoint (direct), which would be wrong IMHO.

So you should check for null, before setting the endpoint

**Comment by bvahdat** [comment]
Thanks, I was observing the behaviour using `3.7.4`. Maybe it has changed since then on the master branch.

**Comment by bvahdat** [comment]
> Thanks, I was observing the behaviour using `3.7.4`. Maybe it has changed since then on the master branch.

Sorry for my stupid comment, now I've got your feedback. Applied as requested.

**Comment by yasserzamani** [comment]
I have doubts about this: Maybe it makes sense to keep it as before as is. I'm not sure maybe someone needs to distinguish between an exchange which has come from a consumer and an exchange which hasn't. With these changes she/he misses this possibility. Furthermore now that I re-think I personally think `null` is rational and means that the exchange is originated by an unknown method and has sent to the first `from`. But without it we cannot distinguish between if it's originated by the first `from` or no. At bottom you know better. That's just my feelings.

**Comment by yasserzamani** [comment]
Another possibility is adding a method which wraps following logic (to keep backward-compatibility as well as addressing above issue that I said):
```java
String getFromRouteEndpointUri()
{
  String fromEndpointUri = null;
  if (message.getExchange().getFromEndpoint() != null) {
      fromEndpointUri = message.getExchange().getFromEndpoint().getEndpointUri();
  } else {
      Route fromRoute = message.getExchange().getContext().getRoute(message.getExchange().getFromRouteId());
      if (fromRoute != null) {
          fromEndpointUri = fromRoute.getEndpoint().getEndpointUri();
      }
  }
}
```

**Comment by davsclaus** [comment]
Yeah lets give this some more thought. I do think there are valid use-cases where an exchange was created outside a route/uow and have the from endpoint set, but not a route id.

What we could do is maybe to make the javadoc of the getter methods on the Exchange interface updated with more details about such situations.

**Comment by bvahdat** [comment]
>Yeah lets give this some more thought. I do think there are valid use-cases where an exchange was created outside a route/uow and have the from endpoint set, but not a route id.

I think you mean vice versa, that's `fromEndpoint` property in _not_ set but `fromRouteId`property _is_ set, right? Would you please review https://github.com/apache/camel/pull/5577?

**Comment by davsclaus** [comment]
Replaced by
https://github.com/apache/camel/pull/5577

Closing this PR

**Pull request closed without merging**